command/command.go

```diff
@@ -1,129 +1,113 @@
 package command
 
 import (
-	"errors"
+	"github.com/bitrise-io/go-utils/env"
 	"io"
-	"os"
 	"os/exec"
 	"strconv"
 	"strings"
 )
 
-// ----------
-
-// Model ...
-type Model struct {
-	cmd *exec.Cmd
+// Opts ...
+type Opts struct {
+	Stdout io.Writer
+	Stderr io.Writer
+	Env    []string
+	Dir    string
 }
 
-// New ...
-func New(name string, args ...string) *Model {
-	return &Model{
-		cmd: exec.Command(name, args...),
-	}
+// Factory ...
+type Factory interface {
+	Create(name string, args []string, opts *Opts) Command
 }
 
-// NewWithStandardOuts - same as NewCommand, but sets the command's
-// stdout and stderr to the standard (OS) out (os.Stdout) and err (os.Stderr)
-func NewWithStandardOuts(name string, args ...string) *Model {
-	return New(name, args...).SetStdout(os.Stdout).SetStderr(os.Stderr)
+type defaultFactory struct {
+	envRepository env.Repository
 }
 
-// NewWithParams ...
-func NewWithParams(params ...string) (*Model, error) {
-	if len(params) == 0 {
-		return nil, errors.New("no command provided")
-	} else if len(params) == 1 {
-		return New(params[0]), nil
-	}
-
-	return New(params[0], params[1:]...), nil
+// NewFactory ...
+func NewFactory(envRepository env.Repository) Factory {
+	return defaultFactory{envRepository: envRepository}
 }
 
-// NewFromSlice ...
-func NewFromSlice(slice []string) (*Model, error) {
-	return NewWithParams(slice...)
-}
-
-// NewWithCmd ...
-func NewWithCmd(cmd *exec.Cmd) *Model {
-	return &Model{
-		cmd: cmd,
+// Create ...
+func (d defaultFactory) Create(name string, args []string, opts *Opts) Command {
```

Comment:
Nit: I'd call this receiver `f` because `d` is confusing at the call site (I always have to look it up).